src/formula/callable_objects.cpp

```diff
@@ -87,8 +87,10 @@ variant attack_type_callable::get_value(const std::string& key) const
 		return variant(att_->id());
 	} else if(key == "description") {
 		return variant(att_->name());
-	} else if(key == "type") {
+	} else if(key == "original_type") {
 		return variant(att_->type());
+	} else if(key == "type") {
+		return variant(att_->effective_damage_type().first);
```

Comment:
@CelticMinstrel any concerns about changing what `type` refers to here?

Reply:
It definitely looks concerning. I have no idea what `effective_damage_type()` is doing but it's not the same as what `type()` returns, so the correct answer is probably to leave `"type"` alone here and add an `"effective_type"` that uses the new function.

Reply:
effective_damage_type() return the type used in attack same if modified by use of [damage_type], but i will apply your suggest

Reply:
@CelticMinstrel the problem is that "type" in attack_type refers to the type used, including after modification by [damage_type]; it therefore seems logical to me to extend this to the formula and use "original_type" to refer to the type not modified by [damage_type], either I apply your suggestion, and in this case it implies a regression for [filter_attack], or I give up original type, and only filter or checked in the formula the type actually used, but not say 'type' refers to the damage type actually used in the filter but only refers to the original type in the formula

Reply:
I don't really understand what you're saying. The `type()` function just returns the string stored in the class, which presumably is the type set from WML in the attack definition. Changing it to account for `[damage_type]` specials seems like it would be more confusing than useful, as that leads to infinite recursion in some circumstances – although I recall measures being taken against that, breaking the recursion still means the outcome is not well-defined. And changing the meaning of existing formulas also doesn't sound like a great idea.

But we can discuss this. There may be a third option too, or even fourth or fifth options.

> either I apply your suggestion, and in this case it implies a regression for [filter_attack]

Please explain the regression that would occur if you apply my suggestion.

> or I give up original type, and only filter or checked in the formula the type actually used, but not say 'type' refers to the damage type actually used in the filter but only refers to the original type in the formula

I'm really trying hard to parse this run-on sentence. From what I can tell, you are suggesting to just undo this part of the diff and leave it how it previously was, and also adding that a clarification should be made to the documentation?

I think there _is_ some benefit to having both versions of `type`, so that doesn't seem like a great option. However, I think it is not at all obvious whether it makes more sense for `type` to be the original type or the effective type. That's why I think it's better to leave `type` unchanged.

-----

By the way, as an aside: whatever we ultimately decide on here, the same change should also be applied in the Lua API.

Reply:
i will remove all modifications in formula in regard of infinite recursion issue. For LUA API, i cannot apply change, i already try for attack_alignment attribute and it never worked, because data of units(self or opponent) can't be added in attack attribute in lua, and it will the same for [damage_type] weapon

Reply:
@CelticMinstrel [filter_attack]type= already takes into account the modifications by [damage_type], except that it integrates at the same time the original type (or replace by [damage_type]replacement_type=) and the alternative type ([damage_type]alternative_type=), here, I want that only the type of damage which does the most damage to the opponent is taken into account and appears in the pre-combat window, as well as for the check by [filter_attack/weapon]

Reply:
So in summary, you're saying that your goal is to make the WFL API behave more like as the filter API with respect to the meaning of `type`? Keeping them both the same does make some sense, but… you also seem to be saying you want it to be _different_ from how `[filter_attack]type=` works. I find that difficult to reconcile. It seems like there might be three different possible versions of `type` here?

Reply:
@CelticMinstrel  I hesitate because of the infinite recursion if the formula is used from a [damage_type] affecting the checked attack, but if I modify the code of the formula, it is good for the attributes of filter_weapon and the formula to be identical. So, do I modify formula or not?

Reply:
I put the last commit back in place